### ISSUE TITLE
Feature Request: each window it's own recent tab

### ISSUE THREAD

**Issue by alber70g** [fwextensions/QuicKey#50]
For me it's more logical to switch between recent tabs within one window. So each window should have it's own recent tab. 

If I'd want to switch between chrome windows I would use ALT+TAB for that. 

I suggest ALT+Q should only switch to the recent tab _within_ the active window

**Comment by fwextensions** [comment]
Thanks for the suggestion.  Your windows must be much more organized than mine!  I tend to be jumping back and forth between windows a lot, just because of where tabs happened to get opened up.  

Are you thinking that just the recents list should be limited to the current window, or search results as well?  

One complication is that if you switched to a window that you haven't used in a while, the recents list would be empty, since the only recent tab for that window would be the current one, and the current tab isn't shown in the list .  

There is an existing option to mark tabs that aren't in the current window with an icon, though that doesn't help with filtering them.

**Comment by alber70g** [comment]
> Thanks for the suggestion. Your windows must be much more organized than mine! I tend to be jumping back and forth between windows a lot, just because of where tabs happened to get opened up.

Like I said, my window management is done with Alt+Tab

> Are you thinking that just the recents list should be limited to the current window, or search results as well?

I'm only using the "recent tab switcher" feature of this plugin (sorry 😇) but I'd say that _all tabs_ should be in the search results.

> One complication is that if you switched to a window that you haven't used in a while, the recents list would be empty, since the only recent tab for that window would be the current one, and the current tab isn't shown in the list .

For me this is no problem. I think it's even a given when starting chrome (I'm using "On Starting" "Continue where you left off")

> There is an existing option to mark tabs that aren't in the current window with an icon, though that doesn't help with filtering them.

This doesn't apply to the "Switch instantly between the two most recent tabs"-option

This is how I use the extension :) (I used to use this one https://chrome.google.com/webstore/detail/alt-%20-q-switch-recent-act/odhjcgnlbagjllfbilicalpigimhdcll?hl=en-GB)

![image](https://user-images.githubusercontent.com/516972/109409931-f5c02e00-7996-11eb-9d28-c08cba5f1cdf.png)

Also, there seems to be a slight delay before the tab is actually switched, not sure where that comes from. Especially when switching between two tabs rapidly (I do that often when I want to read something and write about it in another tab). This used to work with the other extension.

**Comment by fwextensions** [comment]
> > One complication is that if you switched to a window that you haven't used in a while, the recents list would be empty, since the only recent tab for that window would be the current one, and the current tab isn't shown in the list .
>
> For me this is no problem. I think it's even a given when starting chrome (I'm using "On Starting" "Continue where you left off")

The recents list does persist across browser restarts, so you should be able to immediately toggle to the previous tab after relaunching Chrome.

> Also, there seems to be a slight delay before the tab is actually switched, not sure where that comes from. Especially when switching between two tabs rapidly (I do that often when I want to read something and write about it in another tab). 

Hmm, that shouldn't happen.  The time to switch tabs should be basically the same as using ctrl-tab to navigate between them.  Is Chrome overall reacting slowly to keyboard shortcuts?  If you're writing in one tab, could it be that that tab is trying to save something when it loses focus, causing Chrome to slow down?  Maybe try switching between two read-only tabs? 

Here's a screen recording of me rapidly pressing the key to toggle between two tabs:

![K0XHxVaRmO](https://user-images.githubusercontent.com/61631/109444880-1ddd7a80-79f3-11eb-9744-478ed0413b53.gif)

**Comment by alber70g** [comment]
Either way, is it possible to make the recent-tab-switch go “per-window” instead of “globally”?

**Comment by srsudar** [comment]
I use the [Previous Tab](https://chrome.google.com/webstore/detail/previous-tab/bjaniflnlhhofabpoamhnobeonjcjjpl) extension to do last tab per-window. I'm not sure how it does it, as iirc it doesn't require any big permissions.

**Comment by fwextensions** [comment]
Yes, it's certainly possible to limit the list of recent tabs to the current window.  The window ID is stored with each tab ID, since QuicKey needs that to focus the window and then the tab when jumping between windows.  

When using the keyboard shortcut to navigate back into the previous tab stack, I guess the call could keep going until it found a previous tab with the same window ID, or stop if it looped around.  So it may not be too hard to do.

**Comment by carlocardella** [comment]
Limiting recent-tab-switch to the current window could be useful in certain scenarios but I think the behavior should be controlled by a setting one can choose to enable. I personally prefer the current behavior for example: I try to keep my windows organized per topic but sometimes I need to switch between topics/windows on purpose and the current behavior is exactly what I need
